File: 0348. Design Tic-Tac-Toe/Design Tic-Tac-Toe.cpp
class TicTacToe {
    vector<int> rows, cols;
    int diagonal, rdiagonal;
public:
    /** Initialize your data structure here. */
    TicTacToe(int n) {
        rows = vector<int>(n, 0);
        cols = vector<int>(n, 0);
        diagonal = 0;
        rdiagonal = 0;
    }
    
    /** Player {player} makes a move at ({row}, {col}).
        @param row The row of the board.
        @param col The column of the board.
        @param player The player, can be either 1 or 2.
        @return The current winning condition, can be either:
                0: No one wins.
                1: Player 1 wins.
                2: Player 2 wins. */
    int move(int row, int col, int player) {
        int toAdd = player == 1 ? 1 : -1;
        rows[row] += toAdd;
        cols[col] += toAdd;
        if (row == col) diagonal += toAdd;
        if (row + col == rows.size() - 1) rdiagonal += toAdd;
        
        int s = rows.size();
        if (abs(rows[row]) == s || abs(cols[col]) == s ||
            abs(diagonal) == s || abs(rdiagonal) == s) return player;
        return 0;
    }
};

/**
 * Your TicTacToe object will be instantiated and called as such:
 * TicTacToe* obj = new TicTacToe(n);
 * int param_1 = obj->move(row,col,player);
 */

File: 0348. Design Tic-Tac-Toe/Design Tic-Tac-Toe.py
class TicTacToe:

    def __init__(self, n: int):
        """
        Initialize your data structure here.
        """
        self.rows, self.cols = [0] * n, [0] * n
        self.diagonal, self.rdiagonal = 0, 0
        

    def move(self, row: int, col: int, player: int) -> int:
        """
        Player {player} makes a move at ({row}, {col}).
        @param row The row of the board.
        @param col The column of the board.
        @param player The player, can be either 1 or 2.
        @return The current winning condition, can be either:
                0: No one wins.
                1: Player 1 wins.
                2: Player 2 wins.
        """
        toAdd = 1 if player == 1 else -1
        
        self.rows[row], self.cols[col] = self.rows[row] + toAdd, self.cols[col] + toAdd
        if row == col:
            self.diagonal += toAdd
        
        if row + col == len(self.rows) - 1:
            self.rdiagonal += toAdd
            
        s = len(self.rows)
        return player if abs(self.rows[row]) == s or \
                         abs(self.cols[col]) == s or \
                         abs(self.diagonal) == s or \
                         abs(self.rdiagonal) == s else 0

# Your TicTacToe object will be instantiated and called as such:
# obj = TicTacToe(n)
# param_1 = obj.move(row,col,player)